app/controllers/openid_connect/authorization_controller.rb

```diff
@@ -84,10 +105,26 @@ def session_params
       user_session.delete(:openid_auth_request) || {}
     end
 
+    def store_request
+      return if sp_session[:request_id]
+
+      @request_id = SecureRandom.uuid
+      ServiceProviderRequest.find_or_create_by(uuid: @request_id) do |sp_request|
+        sp_request.issuer = @authorize_form.client_id
+        sp_request.loa = @authorize_form.acr_values.sort.max
+        sp_request.url = request.original_url
+      end
+    end
+
     def add_sp_metadata_to_session
-      session[:sp] = { loa3: @authorize_form.loa3_requested?,
-                       issuer: @authorize_form.client_id,
-                       show_start_page: true }
+      return if sp_session[:request_id]
+
+      session[:sp] = {
```

Comment:
I think @monfresh has taken the right approach. Populating the session once from the db is cleaner, since there is already a single link between the session and the browser (the session cookie).